src/lightning/fabric/plugins/precision/bitsandbytes.py

```diff
@@ -192,51 +221,160 @@ class _Linear8bitLt(bnb.nn.Linear8bitLt):
+            if self.bias is not None:
+                fan_in, _ = torch.nn.init._calculate_fan_in_and_fan_out(self.weight)
+                bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
+                init.uniform_(self.bias, -bound, bound)
+
+            linear_init_finished = isinstance(self.weight, bnb.nn.Params4bit)
+            if linear_init_finished and self.weight.dtype == torch.uint8:  # was quantized
+                # need the original shape here
+                raise NotImplementedError
+            weight = self.weight.data
+            torch.nn.init.kaiming_uniform_(weight, a=math.sqrt(5))
+            if linear_init_finished:
+                if self.weight.device.type == "meta":
+                    # need custom logic if int8params is on meta device
+                    raise NotImplementedError
+                if self.weight.device.type == "cuda":  # re-quantize
+                    self.quantize_(weight)
+                else:
+                    self.weight = _replace_param(self.weight, weight)
 
     class _Linear4bit(bnb.nn.Linear4bit):
-        """Wraps `bnb.nn.Linear4bit` and enables instantiation directly on the device and re-quantizaton when loading
-        the state dict."""
+        """Wraps `bnb.nn.Linear4bit` to enable: instantiation directly on the device, re-quantizaton when loading the
+        state dict, meta-device initialization, and materialization."""
 
         def __init__(self, *args: Any, device: Optional[_DEVICE] = None, **kwargs: Any) -> None:
             super().__init__(*args, device=device, **kwargs)
+            self.weight = cast(bnb.nn.Params4bit, self.weight)  # type: ignore[has-type]
+            self.bias = cast(Optional[torch.nn.Parameter], self.bias)  # type: ignore[has-type]
             # if the device is CUDA or we are under a CUDA context manager, quantize the weight here, so we don't end up
             # filling the device memory with float32 weights which could lead to OOM
             if torch.tensor(0, device=device).device.type == "cuda":
-                self._quantize_weight(self.weight.data)
-            self._register_load_state_dict_pre_hook(partial(_quantize_on_load_hook, self._quantize_weight))
+                self.quantize_()
+            self._register_load_state_dict_pre_hook(partial(_quantize_on_load_hook, self.quantize_))
             self.register_load_state_dict_post_hook(_ignore_missing_weights_hook)
 
-        def _quantize_weight(self, weight: torch.Tensor) -> None:
+        def quantize_(self, weight: Optional[torch.Tensor] = None, device: Optional[torch.device] = None) -> None:
+            """Inplace quantize."""
+            if weight is None:
+                weight = self.weight.data
+                if weight.data.type == torch.uint8:
+                    # already quantized
+                    return
+            assert isinstance(self.weight, bnb.nn.Params4bit)
+            self.weight = self.quantize(self.weight, weight, device)
```

Comment:
looks like the same as above, can we share it?

Reply:
Not really because that would require using a mixin since there's no shared interface other than `torch.nn.Linear`